tgui/packages/tgui/interfaces/ControllerOverview.tsx

```diff
@@ -0,0 +1,228 @@
+  } else if (data.doesnt_fire) {
+    icon = 'check';
+  } else if (!data.can_fire) {
+    icon = 'pause';
+  }
+
+  return (
+    <Collapsible
+      title={name}
+      key={data.ref}
+      icon={icon}
+      buttons={
+        <Button
+          icon="wrench"
+          tooltip="View Variables"
+          onClick={() => {
+            act('view_variables', { ref: data.ref });
+          }}
+        />
+      }
+    >
+      <Stack vertical>
+        <Stack.Item>Init Order: {data.init_order}</Stack.Item>
+        <Stack.Item>Last Fire: {data.last_fire}</Stack.Item>
+        <Stack.Item>Next Fire: {data.next_fire}</Stack.Item>
+        <Stack.Item>Cost: {data.cost_ms}ms</Stack.Item>
+        <Stack.Item>
+          Tick Usage: {(data.tick_usage * 0.01).toFixed(2)}%
+        </Stack.Item>
+        <Stack.Item>
+          Tick Overrun: {(data.tick_overrun * 0.01).toFixed(2)}%
+        </Stack.Item>
+        {data.initialization_failure_message ? (
+          <Stack.Item color="bad">
+            {data.initialization_failure_message}
+          </Stack.Item>
+        ) : undefined}
+      </Stack>
+    </Collapsible>
+  );
+};
+
+enum SubsystemSortBy {
+  INIT_ORDER = 'Init Order',
+  NAME = 'Name',
+  LAST_FIRE = 'Last Fire',
+  NEXT_FIRE = 'Next Fire',
+  TICK_USAGE = 'Tick Usage',
+  TICK_OVERRUN = 'Tick Overrun',
+}
+
+const sortSubsystemBy = (
+  subsystems: Record<string, SubsystemData>,
+  sortBy: SubsystemSortBy,
+  asending: boolean = true,
+) => {
+  let sorted = Object.entries(subsystems).sort(
+    ([nameA, subsystemA], [nameB, subsystemB]) => {
+      switch (sortBy) {
+        case SubsystemSortBy.INIT_ORDER:
+          return subsystemA.init_order - subsystemB.init_order;
+        case SubsystemSortBy.NAME:
+          return nameA.localeCompare(nameB);
+        case SubsystemSortBy.LAST_FIRE:
+          return subsystemA.last_fire - subsystemB.last_fire;
+        case SubsystemSortBy.NEXT_FIRE:
+          return subsystemA.next_fire - subsystemB.next_fire;
+        case SubsystemSortBy.TICK_USAGE:
+          return subsystemA.tick_usage - subsystemB.tick_usage;
+        case SubsystemSortBy.TICK_OVERRUN:
+          return subsystemA.tick_overrun - subsystemB.tick_overrun;
+      }
+    },
+  );
+  if (!asending) {
+    sorted = sorted.reverse();
+  }
+  return Object.fromEntries(sorted);
+};
+
+export const ControllerOverview = () => {
+  const { act, data } = useBackend<ControllerData>();
+  const { world_time, map_cpu, subsystems } = data;
+
+  const [filterName, setFilterName] = useState<string | undefined>(undefined);
+  const [sortBy, setSortBy] = useState<SubsystemSortBy>(SubsystemSortBy.NAME);
+  const [sortAscending, setSortAscending] = useState<boolean>(true);
+
+  let filteredSubsystems = subsystems;
+  if (filterName !== undefined) {
+    filteredSubsystems = Object.fromEntries(
+      Object.entries(subsystems).filter(([name, subsystem]) =>
+        name.toLowerCase().includes(filterName.toLowerCase()),
```

Comment:
Please use createSearch here instead, and perhaps wrap filteredSubsystems in a useMemo with filterName as the dependency, otherwise this is refiltered on every render

Reply:
```tsx
  const filteredSearch = useMemo(() => {
    if (!filterName) return subsystems;

    return Object.fromEntries(
      Object.entries(subsystems).filter(
        createSearch(filterName, ([name]) => name),
      ),
    );
  }, [filterName]);
  ```
  (maybe)